### PR TITLE
Move text edit to gui

### PR DESCRIPTION
**Pull request by EnQs**

### CODE REVIEW

File: src/EnQs/backend/BackendService.java [hunk 1]
```diff
@@ -1,127 +1,104 @@
...
                 break;
             case MEM_READ:
-                BigDecimal readValue = memoryValue.stripTrailingZeros();
-                readValue = readValue.setScale(readValue.scale() > 0 ? readValue.scale() + 1 : readValue.scale(), RoundingMode.UNNECESSARY);
-                currentValue = readValue;
-                break;
+                return memoryValue;
         }
+        return value;
     }
```
Comment (EnQs) [inline src/EnQs/backend/BackendService.java]: Just comment to comment thing...

File: src/EnQs/backend/BackendService.java [hunk 2]
```diff
@@ -1,127 +1,104 @@
...
+                memoryValue = memoryValue.add(value);
                 break;
             case MEM_READ:
-                BigDecimal readValue = memoryValue.stripTrailingZeros();
-                readValue = readValue.setScale(readValue.scale() > 0 ? readValue.scale() + 1 : readValue.scale(), RoundingMode.UNNECESSARY);
-                currentValue = readValue;
-                break;
+                return memoryValue;
         }
+        return value;
```
Comment (EnQs) [inline src/EnQs/backend/BackendService.java]: Just learning how to comment section of code

File: src/EnQs/gui/DisplayBoard.java [hunk 3]
```diff
@@ -9,38 +10,42 @@
...
 import java.awt.*;
 import java.math.BigDecimal;
+import java.util.EnumMap;
+import java.util.function.Consumer;
 
 public class DisplayBoard extends FontUtility {
+    @SuppressWarnings("FieldCanBeLocal")
     private final String FONT_NAME = Font.MONOSPACED;
+    @SuppressWarnings("FieldCanBeLocal")
     private final int FONT_TYPE = Font.BOLD;
```
Comment (Tyvrel) [inline src/EnQs/gui/DisplayBoard.java]: Use `private static final` and removed warning suppression

File: src/EnQs/backend/BackendController.java [hunk 4]
```diff
@@ -18,15 +18,15 @@ private BackendService buildBackendController() {
...
-        ResultData computationResult = backendService.compute(activationKey);
+    private void getResult(EventData data) {
+        EventData computationResult = backendService.compute(data);
         notifyListener(computationResult);
     }
 
-    private void notifyListener(ResultData computationResult) {
-        ListenerService.notifyResultDataListeners(ListenerType.BACKEND_OUTPUT, computationResult);
+    private void notifyListener(EventData computationResult) {
+        ListenerService.notifyListeners(ListenerType.BACKEND_OUTPUT, computationResult);
```
Comment (Tyvrel) [inline src/EnQs/backend/BackendController.java]: `notifyListener` can be inlined

File: src/EnQs/backend/BackendService.java [hunk 5]
```diff
@@ -1,127 +1,104 @@
...
 import EnQs.gui.KeyProperties;
 
 import java.math.BigDecimal;
-import java.math.BigInteger;
 import java.math.RoundingMode;
 import java.util.HashMap;
-import java.util.function.Consumer;
+import java.util.function.BiFunction;
 
 class BackendService {
```
Comment (Tyvrel) [inline src/EnQs/backend/BackendService.java]: Lower access scope whereever possible

File: src/EnQs/backend/BackendService.java [hunk 6]
```diff
@@ -1,127 +1,104 @@
...
     }
 
-    private ResultData composeResults(KeyProperties activationKey) {
-        boolean unnecessaryTrailingZero = currentValue.scale() > 0 && currentValue.unscaledValue().mod(BigInteger.TEN).equals(BigInteger.ZERO);
-        BigDecimal value = unnecessaryTrailingZero ? currentValue.setScale(currentValue.scale() - 1, RoundingMode.HALF_UP) : currentValue;
-        boolean isDotKeyActivated = activationKey.name().equals(KeyProperties.COMMA.name()) && value.scale() == 0;
-        boolean isAnyKeyActivated = activationKey.getActionType() == ActionType.MATHEMATICAL || isDotKeyActivated;
-        KeyProperties activeKey = isAnyKeyActivated ? activationKey : null;
-        return new ResultData(value, activeKey);
+    private EventData composeResults(BigDecimal result) {
```
Comment (Tyvrel) [inline src/EnQs/backend/BackendService.java]: buildResult

File: src/EnQs/backend/BackendService.java [hunk 7]
```diff
@@ -1,127 +1,104 @@
...
+        keyReactionHashMap.put(ActionType.OPERATION, this::modifyValue);
+        keyReactionHashMap.put(ActionType.MEMORY, this::modifyMemory);
+        keyReactionHashMap.put(ActionType.SPECIAL, this::choseSpecialFunction);
+        return keyReactionHashMap;
     }
 
-    private void doMathematicalOperation(KeyProperties activationKey) {
-        lockedValue = BiFunctionType.valueOf(selectedOperation).apply(lockedValue, currentValue).stripTrailingZeros();
+    private BigDecimal doMathematicalOperation(KeyProperties activationKey, BigDecimal value) {
+        String operation = selectedOperation == null ? KeyProperties.PLUS.name() : selectedOperation;
```
Comment (Tyvrel) [inline src/EnQs/backend/BackendService.java]: Why?
If selectedOperation is null, shouldn't you just set "value" as "lockedValue"?

File: src/EnQs/backend/BackendService.java [hunk 8]
```diff
@@ -1,127 +1,104 @@
...
-import java.math.BigInteger;
 import java.math.RoundingMode;
 import java.util.HashMap;
-import java.util.function.Consumer;
+import java.util.function.BiFunction;
 
 class BackendService {
 
-    private final HashMap<ActionType, Consumer<KeyProperties>> keyReactionHashMap;
+    private final HashMap<ActionType, BiFunction<KeyProperties, BigDecimal, BigDecimal>> keyReactionHashMap;
```
Comment (Tyvrel) [inline src/EnQs/backend/BackendService.java]: rename `functionByActionType` or `functionByActionTypeMap`

File: src/EnQs/core/listenerService/ListenerService.java [hunk 9]
```diff
@@ -9,36 +8,21 @@
...
 
-    private static HashMap<ListenerType, ArrayList<Consumer<KeyProperties>>> keyPropertiesListeners;
-    private static HashMap<ListenerType, ArrayList<Consumer<ResultData>>> resultDataListeners;
+    private static HashMap<ListenerType, ArrayList<Consumer<EventData>>> listenersList;
 
     public ListenerService() {
-        keyPropertiesListeners = new HashMap<>();
-        resultDataListeners = new HashMap<>();
+        listenersList = new HashMap<>();
     }
```
Comment (Tyvrel) [inline src/EnQs/core/listenerService/ListenerService.java]: ↑↑↑ Do not initialize static fields in constructor!
`private static HashMap<ListenerType, ArrayList<Consumer<EventData>>> listenersList = new HashMap<>();`

File: src/EnQs/core/listenerService/ListenerService.java [hunk 10]
```diff
@@ -9,36 +8,21 @@
 
 public class ListenerService {
 
-    private static HashMap<ListenerType, ArrayList<Consumer<KeyProperties>>> keyPropertiesListeners;
-    private static HashMap<ListenerType, ArrayList<Consumer<ResultData>>> resultDataListeners;
+    private static HashMap<ListenerType, ArrayList<Consumer<EventData>>> listenersList;
 
     public ListenerService() {
```
Comment (Tyvrel) [inline src/EnQs/core/listenerService/ListenerService.java]: Do not create ListenerService instance at all

File: src/EnQs/gui/DisplayBoard.java [hunk 11]
```diff
@@ -9,38 +10,42 @@
...
 
     private final JPanel displayPanel;
     private final TLabel numericalDisplay;
 
+    private String displayedValue;
+    private boolean isDisplayedFinal;
```
Comment (Tyvrel) [inline src/EnQs/gui/DisplayBoard.java]: What problem `isDisplayedFinal` is trying to fix? It is almost always "false"

File: src/EnQs/gui/Keyboard.java [hunk 12]
```diff
@@ -13,23 +13,27 @@
...
             {KeyProperties.MEM_CLEAR, KeyProperties.MEM_READ, KeyProperties.MEM_PLUS, KeyProperties.BACKSPACE, KeyProperties.ALL_CLEAR},
-            {KeyProperties.SEVEN, KeyProperties.EIGHT, KeyProperties.NINE, KeyProperties.PERCENT, KeyProperties.SQRT},
+            {KeyProperties.SEVEN, KeyProperties.EIGHT, KeyProperties.NINE, KeyProperties.POWER, KeyProperties.SQRT},
             {KeyProperties.FOUR, KeyProperties.FIVE, KeyProperties.SIX, KeyProperties.MULTIPLY, KeyProperties.DIVIDE},
             {KeyProperties.ONE, KeyProperties.TWO, KeyProperties.THREE, KeyProperties.PLUS, KeyProperties.MINUS},
-            {KeyProperties.SIGN, KeyProperties.ZERO, KeyProperties.COMMA, KeyProperties.POWER, KeyProperties.EQUAL}
+            {KeyProperties.COMMA, KeyProperties.ZERO, KeyProperties.SIGN, KeyProperties.PERCENT, KeyProperties.EQUAL}
     };
 
     private final Dimension KEYPAD;
```
Comment (Tyvrel) [inline src/EnQs/gui/Keyboard.java]: `KEYPAD` should be local variable

File: src/EnQs/backend/BackendService.java (hunk 5 above)
Comment (EnQs) [reply src/EnQs/backend/BackendService.java]: I don't understand:

> Lower access scope

File: src/EnQs/backend/BackendService.java (hunk 7 above)
Comment (EnQs) [reply src/EnQs/backend/BackendService.java]: `selectedOperation  == null` means that operational memory was cleared by either `EQUAL` or `AC` buttons, that means also `lockedValue == 0` and both fields await for new input.
If bellow code looks to hacky: 
```java
private BigDecimal doMathematicalOperation(KeyProperties activationKey, BigDecimal value) {
        String operation = selectedOperation == null ? KeyProperties.PLUS.name() : selectedOperation;
```
I can will this to:
```java
if (selectedOperation  == null) {
        lockedValue = value
}else {
        lockedValue = BiFunctionType.valueOf(operation).apply(lockedValue, value);
}
```
Is that what you mean?

File: src/EnQs/gui/DisplayBoard.java (hunk 11 above)
Comment (EnQs) [reply src/EnQs/gui/DisplayBoard.java]: We will discuss that later

**Pull request merged**